File: shipbob_to_gorgias/java/src/main/java/sb2gorgias/GorgiasApi.java
package sb2gorgias;

import java.time.Duration;
import java.time.format.DateTimeFormatter;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

import org.json.JSONArray;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** Who a ShipBob order or shipment event ships to. Both flows key their Gorgias customer off
 * this. */
record Recipient(String name, String email, Address address) {
}

record Address(String address1, String city, String country) {
}

/** Gorgias API client.
 *
 * The cron flow upserts customers (writing ShipBob order history to
 * data.pandium.shipbob_orders); the webhook flow creates tickets.
 *
 * Auth is OAuth2 via Pandium's gorgias-oauth connector. Pandium runs the authorization flow
 * when the tenant connects and refreshes the token on its own schedule, so this client never
 * sees a client secret, never posts to a token endpoint, and holds no refresh logic - it reads
 * whatever access token is current for this run and sends it as a bearer token. A refresh that
 * fails is a platform concern and surfaces as Failed (Refresh) on the run, not as an error this
 * code has to handle. */
final class GorgiasApi implements GorgiasClient {
    private static final Logger LOGGER = LoggerFactory.getLogger("gorgias");

    // Only a recipient email Gorgias would actually accept counts as valid.
    private static final Pattern EMAIL_RE = Pattern.compile(
            "([-!#-'*+/-9=?A-Z^-~]+(\\.[-!#-'*+/-9=?A-Z^-~]+)*|\"([\\]!#-\\[^-~ \\t]|(\\\\[\\t -~]))+\")"
            + "@([-!#-'*+/-9=?A-Z^-~]+(\\.[-!#-'*+/-9=?A-Z^-~]+)*|\\[[\\t -Z^-~]*])"
    );

    private static final DateTimeFormatter DISPLAY_DATE = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss 'UTC'");

    final String apiUrl;
    final ApiClient apiClient;

    GorgiasApi(Pandium pandium) {
        Map<String, String> secrets = pandium.secrets;
        String token = secrets.get("gorgias_oauth_access_token");
        String account = secrets.get("gorgias_oauth_account");
        if (token == null || token.isEmpty() || account == null || account.isEmpty()) {
            throw new IllegalStateException(
                    "PAN_SEC_GORGIAS_OAUTH_ACCESS_TOKEN and PAN_SEC_GORGIAS_OAUTH_ACCOUNT are required");
        }
        this.apiUrl = "https://" + account.toLowerCase() + ".gorgias.com/api";
        // The connector reports its own scheme; every current Gorgias token is a bearer.
        String tokenType = secrets.get("gorgias_oauth_token_type");
        if (tokenType == null || tokenType.isEmpty()) {
            tokenType = "Bearer";
        }
        // Exponential backoff: 2s, 4s, 8s, ... GET/POST/PUT are all retried.
        this.apiClient = new ApiClient(apiUrl, tokenType + " " + token, Duration.ofSeconds(2),
                Set.of("GET", "POST", "PUT"));
    }

    /** Looks a customer up by email or externalId and returns the detail record (so callers
     * can read data), or null if not found. A given email/externalId maps to at most one
     * customer, so no pagination is needed. */
    @Override
    public JSONObject findCustomer(String email, String externalId) {
        LOGGER.info("looking for gorgias customer: {}, {}", email, externalId);
        Map<String, String> query;
        if (email != null && !email.isEmpty()) {
            query = Map.of("email", email.toLowerCase());
        } else if (externalId != null && !externalId.isEmpty()) {
            query = Map.of("external_id", externalId);
        } else {
            return null;
        }

        Object res = apiClient.get("/customers", query);
        JSONObject body = res instanceof JSONObject j ? j : new JSONObject();
        JSONArray rows = body.optJSONArray("data");
        if (rows == null || rows.isEmpty()) {
            LOGGER.info("customer not found");
            return null;
        }

        JSONObject first = rows.optJSONObject(0);
        if (first == null || !first.has("id") || first.isNull("id")) {
            throw new IllegalStateException("Gorgias customer has no id");
        }
        Object detail = apiClient.get("/customers/" + first.get("id"), null);
        LOGGER.info("customer found");
        return detail instanceof JSONObject j ? j : null;
    }

    @Override
    public long createCustomer(JSONObject payload) {
        LOGGER.info("creating new gorgias customer");
        Object res;
        try {
            res = apiClient.post("/customers", payload);
        } catch (RuntimeException e) {
            LOGGER.error("create customer failed", e);
            throw e;
        }
        if (!(res instanceof JSONObject j) || !j.has("id") || j.isNull("id")) {
            throw new IllegalStateException("Gorgias created a customer without an id");
        }
        LOGGER.info("customer created successfully");
        return j.optLong("id");
    }

    @Override
    public void updateCustomer(long id, JSONObject payload) {
        LOGGER.info("updating gorgias customer {}", id);
        try {
            apiClient.put("/customers/" + id, payload);
        } catch (RuntimeException e) {
            LOGGER.error("update customer {} failed", id, e);
            throw e;
        }
        LOGGER.info("customer updated");
    }

    @Override
    public JSONObject createTicket(JSONObject payload) {
        LOGGER.info("creating gorgias ticket");
        Object res;
        try {
            res = apiClient.post("/tickets", payload);
        } catch (RuntimeException e) {
            LOGGER.error("create ticket failed", e);
            throw e;
        }
        return res instanceof JSONObject j ? j : new JSONObject();
    }

    /** Returns email if Gorgias would accept it, else "". */
    static String validEmail(String email) {
        if (email != null && !email.isEmpty() && !email.contains(".@") && EMAIL_RE.matcher(email).matches()) {
            return email;
        }
        return "";
    }

    /** Reads the recipient common to both a ShipBob order and a shipment event - same shape,
     * same field names - so both flows can key their Gorgias customer off one extraction. */
    static Recipient recipientOf(JSONObject data) {
        JSONObject recipient = data.optJSONObject("recipient");
        if (recipient == null) {
            return new Recipient("", "", new Address("", "", ""));
        }
        JSONObject address = recipient.optJSONObject("address");
        Address addr = address == null ? new Address("", "", "") : new Address(
                address.optString("address1", ""), address.optString("city", ""), address.optString("country", ""));
        return new Recipient(recipient.optString("name", ""), recipient.optString("email", ""), addr);
    }

    /** The key identifying a recipient's customer: a valid email when present, otherwise a
     * synthetic "name address1 city country". */
    static String customerKey(Recipient recipient) {
        String email = validEmail(recipient.email());
        if (!email.isEmpty()) {
            return email;
        }
        Address address = recipient.address();
        return String.join(" ", recipient.name(), address.address1(), address.city(), address.country());
    }

    /** Body for POST /customers when the customer does not yet exist. */
    static JSONObject newCustomerPayload(Recipient recipient, String key) {
        JSONObject payload = new JSONObject();
        payload.put("name", recipient.name());
        payload.put("external_id", key);
        payload.put("data", new JSONObject().put("pandium", new JSONObject().put("shipbob_orders", new JSONArray())));
        String email = validEmail(recipient.email());
        if (!email.isEmpty()) {
            payload.put("email", email);
        }
        return payload;
    }

    /** The single order entry stored in data.pandium.shipbob_orders. */
    static JSONObject orderDataPayload(JSONObject order) {
        Object shipmentsObj = Util.deepGet(order, "shipments", new JSONArray());
        JSONArray shipments = shipmentsObj instanceof JSONArray a ? a : new JSONArray();
        for (int i = 0; i < shipments.length(); i++) {
            if (shipments.opt(i) instanceof JSONObject shipment) {
                for (String field : new String[] {"estimated_fulfillment_date", "actual_fulfillment_date"}) {
                    String value = shipment.optString(field, "");
                    if (!value.isEmpty()) {
                        shipment.put(field, formatDate(value));
                    }
                }
                Object shipmentId = shipment.opt("id");
                shipment.put("url", "https://web.shipbob.com/App/Merchant/#/Orders/"
                        + (shipmentId == null ? "" : shipmentId) + "/");
            }
        }
        JSONObject payload = new JSONObject();
        payload.put("id", Util.deepGet(order, "id", ""));
        payload.put("created_date", formatDate(Util.asString(Util.deepGet(order, "created_date", ""))));
        payload.put("purchase_date", formatDate(Util.asString(Util.deepGet(order, "purchase_date", ""))));
        payload.put("reference_id", Util.deepGet(order, "reference_id", ""));
        payload.put("order_number", Util.deepGet(order, "order_number", ""));
        payload.put("status", Util.deepGet(order, "status", ""));
        payload.put("type", Util.deepGet(order, "type", ""));
        payload.put("channel", Util.deepGet(order, "channel", new JSONObject()));
        payload.put("shipping_method", Util.deepGet(order, "shipping_method", ""));
        payload.put("recipient", Util.deepGet(order, "recipient", new JSONObject()));
        payload.put("products", Util.deepGet(order, "products", new JSONArray()));
        payload.put("tags", Util.deepGet(order, "tags", new JSONArray()));
        payload.put("shipments", shipments);
        return payload;
    }

    /** Renders a ShipBob ISO timestamp for the customer sidebar; passes through anything
     * unparseable. */
    static String formatDate(String value) {
        if (value == null || value.isEmpty()) {
            return "";
        }
        return Util.parseTimestamp(value).map(DISPLAY_DATE::format).orElse(value);
    }
}

File: shipbob_to_gorgias/php/src/GorgiasAPI.php
<?php

declare(strict_types=1);

namespace Pandium\Integration;

use GuzzleHttp\ClientInterface;
use GuzzleHttp\Exception\RequestException;
use Psr\Http\Message\ResponseInterface;
use Psr\Log\LoggerInterface;
use RuntimeException;
use stdClass;
use Throwable;

/**
 * Gorgias API client: the cron flow upserts customers, the webhook flow creates tickets.
 *
 * Auth is OAuth2 via Pandium's `gorgias-oauth` connector. Pandium runs the authorization
 * flow and refreshes the token itself, so this client just sends whatever access token is
 * current for the run; a failed refresh surfaces on the run as **Failed (Refresh)**.
 */
final class GorgiasAPI implements GorgiasClient
{
    /** Mirrors Gorgias's own email validation, so only addresses it accepts key a customer. */
    private const EMAIL_RE = '/^([-!#-\'*+\/-9=?A-Z^-~]+(\.[-!#-\'*+\/-9=?A-Z^-~]+)*'
        . '|"([]!#-[^-~ \t]|(\\\\[\t -~]))+")'
        . '@([-!#-\'*+\/-9=?A-Z^-~]+(\.[-!#-\'*+\/-9=?A-Z^-~]+)*|\[[\t -Z^-~]*])$/D';

    /** ShipBob timestamps are UTC; rearranging the digits avoids any timezone conversion. */
    private const ISO_RE = '/^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})/';

    public readonly string $apiUrl;

    private readonly ClientInterface $http;

    private readonly LoggerInterface $logger;

    /** $handler and $retryDelayMs let tests send through a mock handler without sleeping. */
    public function __construct(Pandium $pandium, ?callable $handler = null, int $retryDelayMs = 2000)
    {
        $token = $pandium->secrets['gorgias_oauth_access_token'] ?? '';
        $account = $pandium->secrets['gorgias_oauth_account'] ?? '';
        if ($token === '' || $account === '') {
            throw new RuntimeException(
                'PAN_SEC_GORGIAS_OAUTH_ACCESS_TOKEN and PAN_SEC_GORGIAS_OAUTH_ACCOUNT are required'
            );
        }

        $this->logger = Log::get('gorgias');
        $this->apiUrl = 'https://' . strtolower($account) . '.gorgias.com/api';

        // The connector reports the token scheme alongside the token.
        $tokenType = $pandium->secrets['gorgias_oauth_token_type'] ?? 'Bearer';
        $this->http = Http::client([
            'accept' => 'application/json',
            'content-type' => 'application/json',
            'Authorization' => "{$tokenType} {$token}",
        ], $retryDelayMs, $handler);
    }

    // --- customers (cron flow) -------------------------------------------------

    /** The full customer record for this email or external_id, or null if none. */
    public function findCustomer(?string $email = null, ?string $externalId = null): ?array
    {
        $this->logger->info("looking for gorgias customer: {$email}, {$externalId}");
        if ($email !== null && $email !== '') {
            $query = ['email' => strtolower($email)];
        } elseif ($externalId !== null && $externalId !== '') {
            $query = ['external_id' => $externalId];
        } else {
            return null;
        }

        $res = $this->http->request('GET', "{$this->apiUrl}/customers", ['query' => $query]);
        $rows = $this->decode($res)['data'] ?? [];
        if ($rows === []) {
            $this->logger->info('Customer not found');

            return null;
        }

        $detail = $this->http->request('GET', "{$this->apiUrl}/customers/{$rows[0]['id']}");
        $this->logger->info('Customer found');

        return $this->decode($detail);
    }

    public function createCustomer(array $payload): int
    {
        $this->logger->info('creating new gorgias customer');
        try {
            $res = $this->http->request('POST', "{$this->apiUrl}/customers", ['json' => $payload]);
        } catch (Throwable $e) {
            $this->logger->error('Create customer failed: ' . self::responseBody($e));

            throw $e;
        }
        $this->logger->info('Customer created successfully');

        return (int) $this->decode($res)['id'];
    }

    public function updateCustomer(int $customerId, array $payload): void
    {
        $this->logger->info("updating gorgias customer {$customerId}");
        try {
            $this->http->request('PUT', "{$this->apiUrl}/customers/{$customerId}", ['json' => $payload]);
        } catch (Throwable $e) {
            $this->logger->error("Update customer {$customerId} failed: " . self::responseBody($e));

            throw $e;
        }
        $this->logger->info('customer updated');
    }

    // --- tickets (webhook flow) ---------------------------------------------------

    public function createTicket(array $payload): array
    {
        $this->logger->info('creating gorgias ticket');
        try {
            $res = $this->http->request('POST', "{$this->apiUrl}/tickets", ['json' => $payload]);
        } catch (Throwable $e) {
            $this->logger->error('Create ticket failed: ' . self::responseBody($e));

            throw $e;
        }

        return $this->decode($res);
    }

    // --- helpers ------------------------------------------------------------

    public function validEmail(?string $email): string
    {
        if ($email !== null && $email !== '' && !str_contains($email, '.@') && preg_match(self::EMAIL_RE, $email)) {
            return $email;
        }

        return '';
    }

    /**
     * The key identifying an order's customer: a valid recipient email when present,
     * otherwise a synthetic `name address1 city country`. Both flows use it.
     *
     * @param array<string, mixed> $sbOrder
     */
    public function customerKey(array $sbOrder): string
    {
        $email = $this->validEmail($sbOrder['recipient']['email'] ?? null);
        if ($email !== '') {
            return $email;
        }
        $address = $sbOrder['recipient']['address'] ?? [];

        return implode(' ', [
            $sbOrder['recipient']['name'] ?? '',
            $address['address1'] ?? '',
            $address['city'] ?? '',
            $address['country'] ?? '',
        ]);
    }

    /** Body for POST /customers when the customer does not yet exist. */
    public function newCustomerPayload(array $sbOrder, string $key): array
    {
        $payload = [
            'name' => $sbOrder['recipient']['name'] ?? '',
            'external_id' => $key,
            'data' => ['pandium' => ['shipbob_orders' => []]],
        ];
        $email = $this->validEmail($sbOrder['recipient']['email'] ?? null);
        if ($email !== '') {
            $payload['email'] = $email;
        }

        return $payload;
    }

    /** The single order entry stored in `data.pandium.shipbob_orders`. */
    public function orderDataPayload(array $sbOrder): array
    {
        $shipments = $sbOrder['shipments'] ?? [];
        foreach ($shipments as &$shipment) {
            foreach (['estimated_fulfillment_date', 'actual_fulfillment_date'] as $field) {
                if (!empty($shipment[$field])) {
                    $shipment[$field] = self::formatDate($shipment[$field]);
                }
            }
            $shipment['url'] = 'https://web.shipbob.com/App/Merchant/#/Orders/'
                . ($shipment['id'] ?? '') . '/';
        }
        unset($shipment); // break the reference the loop leaves behind

        return [
            'id' => $sbOrder['id'] ?? '',
            'created_date' => self::formatDate($sbOrder['created_date'] ?? ''),
            'purchase_date' => self::formatDate($sbOrder['purchase_date'] ?? ''),
            'reference_id' => $sbOrder['reference_id'] ?? '',
            'order_number' => $sbOrder['order_number'] ?? '',
            'status' => $sbOrder['status'] ?? '',
            'type' => $sbOrder['type'] ?? '',
            'channel' => $sbOrder['channel'] ?? new stdClass(),
            'shipping_method' => $sbOrder['shipping_method'] ?? '',
            'recipient' => $sbOrder['recipient'] ?? new stdClass(),
            'products' => $sbOrder['products'] ?? [],
            'tags' => $sbOrder['tags'] ?? [],
            'shipments' => $shipments,
        ];
    }

    /** Render a ShipBob ISO timestamp for the sidebar; pass through anything unparseable. */
    public static function formatDate(string $value): string
    {
        if ($value === '' || !preg_match(self::ISO_RE, $value, $m)) {
            return $value;
        }
        [, $year, $month, $day, $hour, $minute, $second] = $m;

        return "{$day}/{$month}/{$year} {$hour}:{$minute}:{$second} UTC";
    }

    /** @return array<string, mixed> */
    private function decode(ResponseInterface $res): array
    {
        return json_decode((string) $res->getBody(), true, 512, JSON_THROW_ON_ERROR) ?? [];
    }

    /** The response body of a rejected write, where Gorgias puts its validation errors. */
    private static function responseBody(Throwable $e): string
    {
        return $e instanceof RequestException && $e->getResponse() !== null
            ? (string) $e->getResponse()->getBody()
            : $e->getMessage();
    }
}
